# Multiple Linear Regression

In [4]:
import csv
import pandas as pd
import numpy as np
import sklearn
from sklearn import *

/Users/emily-air13/anaconda3/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
/Users/emily-air13/anaconda3/lib/python3.6/site-packages/sklearn/grid_search.py:42: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)
/Users/emily-air13/anaconda3/lib/python3.6/site-packages/sklearn/learning_curve.py:22: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the functions are moved. This module will 

In [5]:
pd.set_option('display.max_columns', 65)

df = pd.read_csv('clean_data.csv', low_memory=False)

Variables in the Cleaned Data:

Location & Zoning Data (Potential Categorical X):
<div style="-webkit-column-count: 2; -moz-column-count: 2; column-count: 2; -webkit-column-rule: 1px dotted #e0e0e0; -moz-column-rule: 1px dotted #e0e0e0; column-rule: 1px dotted #e0e0e0;">
    <div style="display: inline-block;">
        <ul>
         <li>fips</li>
          <li>latitude</li>
        <li>longitude</li>
         <li>propertycountylandusecode</li>
         
         
         </ul>
    </div>
    <div style="display: inline-block;">
        <ul>
        <li>propertylandusetypeid</li>
         <li>rawcensustractandblock</li>
         <li>regionidcity</li>
         <li>zipcode</li>
        </ul>
    </div>
</div>

Categorical X:

<div style="-webkit-column-count: 2; -moz-column-count: 2; column-count: 2; -webkit-column-rule: 1px dotted #e0e0e0; -moz-column-rule: 1px dotted #e0e0e0; column-rule: 1px dotted #e0e0e0;">
    <div style="display: inline-block;">
        <ul>
         <li>airconditioningtypeid</li>
         <li>hashottuborspa</li>
         <li>heatingorsystemtypeid</li>
        <li>fireplaceflag</li>
        <li>assessmentyear</li>
        </ul>
    </div>
    <div style="display: inline-block;">
        <ul>
        <li>taxdelinquencyflag</li>
        <li>setyear</li>
        <li>pooltype</li> 
        <li>transactiondate</li>
        </ul>
    </div>
</div>


Numerical X:
<div style="-webkit-column-count: 2; -moz-column-count: 2; column-count: 2; -webkit-column-rule: 1px dotted #e0e0e0; -moz-column-rule: 1px dotted #e0e0e0; column-rule: 1px dotted #e0e0e0;">
    <div style="display: inline-block;">
        <ul>
        <li>bedroomcnt</li>
        <li>buildingqualitytypeid</li>
        <li>calculatedbathnbr</li>
        <li>calculatedfinishedsquarefeet</li>
        <li>fireplacecnt</li>
        <li>fullbathcnt</li>
        <li>garagecarcnt</li>
        <li>garagetotalsqft</li>
        <li>lotsizesquarefeet</li>
        <li>poolcnt</li>
        </ul>
    </div>
    <div style="display: inline-block;">
        <ul>
        <li>roomcnt</li>
        <li>unitcnt</li>
        <li>yearbuilt</li>
        <li>numberofstories</li>
        <li>structuretaxvaluedollarcnt</li>
        <li>taxvaluedollarcnt</li>
        <li>landtaxvaluedollarcnt</li>
        <li>taxamount</li>
        <li>taxdelinquencyyear TO REMOVE</li> 
        </ul>
    </div>
</div>
           
                        
Y:
* logerror

In [7]:
def hottub(a):
    if a:
        return 1
    else:
        return 0
    
df['hashottuborspa'] = df['hashottuborspa'].map(lambda a: hottub(a))

def fire(a):
    if a:
        return 1
    else:
        return 0

df['fireplaceflag'] = df['fireplaceflag'].map(lambda a: fire(a))

def tax(a):
    if a=='Y':
        return 1
    else:
        return 0

df['taxdelinquencyflag'] = df['taxdelinquencyflag'].map(lambda a: tax(a))



fips = pd.get_dummies(df['fips'])
df = df.join(fips)

yr = pd.get_dummies(df['setyear'])
df = df.join(yr)

ac = pd.get_dummies(df['airconditioningtypeid'], prefix='ac')
df = df.join(ac)

heat = pd.get_dummies(df['heatingorsystemtypeid'], prefix='heat')
df=df.join(heat)

pool = pd.get_dummies(df['pooltype'], prefix='pooltype')
df = df.join(pool)

df.head()


parcelid  airconditioningtypeid  bedroomcnt  buildingqualitytypeid  \
0  17073783                    5.0         3.0               6.015461   
1  17088994                    5.0         2.0               6.015461   
2  17100444                    5.0         3.0               6.015461   
3  17102429                    5.0         2.0               6.015461   
4  17109604                    5.0         4.0               6.015461   

   calculatedbathnbr  calculatedfinishedsquarefeet    fips  fireplacecnt  \
0                2.5                        1264.0  6111.0           0.0   
1                1.0                         777.0  6111.0           0.0   
2                2.0                        1101.0  6111.0           0.0   
3                1.5                        1554.0  6111.0           1.0   
4                2.5                        2415.0  6111.0           1.0   

   fullbathcnt  garagecarcnt  garagetotalsqft  hashottuborspa  \
0          2.0           2.0            434.0               0   
1          1.0           1.0            434.0               0   
2          2.0           2.0            441.0               0   
3          1.0           2.0            460.0               0   
4          2.0           2.0            665.0               0   

   heatingorsystemtypeid   latitude   longitude  lotsizesquarefeet  poolcnt  \
0                    2.0  34.303597 -119.287236             1735.0      0.0   
1                    2.0  34.272866 -119.198911             7200.0      0.0   
2                    2.0  34.340801 -119.079610             6569.0      0.0   
3                    2.0  34.354313 -119.076405             7400.0      0.0   
4                    2.0  34.266578 -119.165392             6326.0      0.0   

  propertycountylandusecode  propertylandusetypeid  rawcensustractandblock  \
0                      1128                  265.0          61110022003007   
1                      1129                  266.0          61110015031002   
2                      1111                  261.0          61110007011007   
3                      1110                  261.0          61110008002013   
4                      1111                  261.0          61110014021007   

   regionidcity  roomcnt  unitcnt  yearbuilt  numberofstories  fireplaceflag  \
0       34543.0      5.0      1.0     1986.0              2.0              0   
1       34543.0      4.0      1.0     1990.0              1.0              0   
2       26965.0      5.0      1.0     1956.0              1.0              0   
3       26965.0      5.0      1.0     1965.0              1.0              1   
4       34543.0      8.0      1.0     1984.0              2.0              1   

   structuretaxvaluedollarcnt  taxvaluedollarcnt  assessmentyear  \
0                    115087.0           191811.0          2015.0   
1                    143809.0           239679.0          2015.0   
2                     33619.0            47853.0          2015.0   
3                     45609.0            62914.0          2015.0   
4                    277000.0           554000.0          2015.0   

   landtaxvaluedollarcnt  taxamount  taxdelinquencyflag  taxdelinquencyyear  \
0                76724.0    2015.06                   0                 NaN   
1                95870.0    2581.30                   0                 NaN   
2                14234.0     591.64                   0                 NaN   
3                17305.0     682.78                   0                 NaN   
4               277000.0    5886.92                   0                 NaN   

   logerror transactiondate  setyear  pooltype  zipcode  6037.0  6059.0  \
0    0.0953      2016-01-27     2016         0    93001       0       0   
1    0.0198      2016-03-30     2016         0    93004       0       0   
2    0.0060      2016-05-27     2016         0    93066       0       0   
3   -0.0566      2016-06-07     2016         0    93060       0       0   
4    0.0573      2016-08-08     

In [8]:
df_linreg = df

In [9]:
msk = np.random.rand(len(df)) < 0.8

In [10]:
train = df[msk]
test = df[~msk]

In [11]:
train.describe()

parcelid  airconditioningtypeid     bedroomcnt  \
count  1.344850e+05          134485.000000  134485.000000   
mean   1.299589e+07               3.976748       3.042726   
std    2.975446e+06               2.245188       1.149448   
min    1.071174e+07               1.000000       0.000000   
25%    1.154790e+07               1.000000       2.000000   
50%    1.254096e+07               5.000000       3.000000   
75%    1.422095e+07               5.000000       4.000000   
max    1.676893e+08              13.000000      16.000000   

       buildingqualitytypeid  calculatedbathnbr  calculatedfinishedsquarefeet  \
count          134485.000000      134485.000000                 134485.000000   
mean                6.013180           2.309852                   1777.083987   
std                 1.504816           0.972148                    937.588005   
min                 1.000000           1.000000                    120.000000   
25%                 6.000000           2.000000                   1185.000000   
50%                 6.015461           2.000000                   1541.000000   
75%                 7.000000           3.000000                   2097.000000   
max                12.000000          15.000000                  35640.000000   

                fips   fireplacecnt    fullbathcnt   garagecarcnt  \
count  134485.000000  134485.000000  134485.000000  134485.000000   
mean     6048.869710       0.127129       2.243328       0.599710   
std        20.742881       0.400576       0.958880       0.919009   
min      6037.000000       0.000000       1.000000       0.000000   
25%      6037.000000       0.000000       2.000000       0.000000   
50%      6037.000000       0.000000       2.000000       0.000000   
75%      6059.000000       0.000000       3.000000       2.000000   
max      6111.000000       5.000000      15.000000      14.000000   

       garagetotalsqft  hashottuborspa  heatingorsystemtypeid       latitude  \
count    134485.000000   134485.000000          134485.000000  134485.000000   
mean        157.392616        0.023148               3.211473      34.007341   
std         237.824164        0.150373               3.046802       0.265468   
min           0.000000        0.000000               1.000000      33.339295   
25%           0.000000        0.000000               2.000000      33.813293   
50%           0.000000        0.000000               2.000000      34.022330   
75%         434.000000        0.000000               2.000000      34.173893   
max        4384.000000        1.000000              24.000000      34.818767   

           longitude  lotsizesquarefeet        poolcnt  propertylandusetypeid  \
count  134485.000000       1.344850e+05  134485.000000          134485.000000   
mean     -118.201373       2.602116e+04       0.202260             261.833074   
std         0.360412       8.586222e+04       0.401687               5.186948   
min      -119.475416       2.360000e+02       0.000000              31.000000   
25%      -118.414000       5.953000e+03       0.000000             261.000000   
50%      -118.176989       7.200000e+03       0.000000             261.000000   
75%      -117.923797       1.046200e+04       0.000000             266.000000   
max      -117.554636       2.831752e+06       1.000000             275.000000   

       rawcensustractandblock   regionidcity        roomcnt        unitcnt  \
count            1.344850e+05  131851.000000  134485.000000  134485.000000   
mean             6.049180e+13   33721.820403       5.580767       1.068558   
std              2.058465e+11   46861.684421       1.970380       0.368702   
min              6.037101e+13    3491.000000       1.000000       1.000000   
25%              6.037312e+13   12447.000000       4.000000       1.000000   
50%              6.037604e+13   25218.000000       5.000000       1.000000   
75%              6.059042e+13   45457.000000       7.000000       1.000000   
max              6.111009e+13  396

### Basic Linear Regression

In [24]:
lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[2,3,4,5,7,8,9,10,11,13,14,15,16,21,22,24,26,27,29,30,25,31,38,39,40,41,42,43,44,45,46,47,48,29,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64]], array_train[:,33])


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [25]:
lm.intercept_

-4884.8425214777526

In [26]:
lm.coef_

array([  1.03907479e-03,  -1.25334157e-03,  -3.31829230e-03,
         1.30569662e-05,   2.37541806e-03,   3.53169552e-03,
         2.71076289e-03,  -2.10907984e-05,   6.99182701e+03,
        -2.20938080e-03,   5.61104018e-03,   3.36917765e-08,
         2.77967360e+03,  -3.28541993e-04,  -9.21071706e-03,
         1.39733762e-03,  -2.73373189e-07,   3.09912309e-07,
        -4.94356619e+05,  -3.52454976e-06,  -1.00915629e-02,
         2.12007790e-02,  -8.26818701e-02,  -7.80795367e-02,
        -6.85218923e-02,   5.05716768e-04,   4.76042263e-03,
        -2.89080948e-02,   8.01456699e-02,  -2.76990736e-02,
        -5.81152265e-02,  -2.58727426e-02,  -3.12418168e-02,
         4.94356619e+05,  -1.92410900e-03,  -2.61898823e-02,
        -2.82455837e-02,  -2.61420390e-02,  -5.96344098e-02,
        -6.64964581e-02,  -3.94477446e-02,  -4.04065321e-02,
        -4.26588892e-02,  -4.27460184e-03,  -4.54700140e-02,
        -4.62124578e-02,   4.88572679e+03,  -4.88579437e+03,
         2.10604474e+03,

In [27]:
array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[2,3,4,5,7,8,9,10,11,13,14,15,16,21,22,24,26,27,29,30,25,31,38,39,40,41,42,43,44,45,46,47,48,29,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64]])

In [28]:
from sklearn.metrics import mean_squared_error
from math import sqrt

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))
 

0.1558109689457013


### Lasso Selection

In [29]:
from sklearn import linear_model
clf = linear_model.Lasso(tol=1) #alpha=0.1
clf.fit(train.iloc[:,[2,3,4,5,7,8,9,10,11,13,14,15,16,21,22,24,26,27,29,30,25,31,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64]], train.iloc[:,33])


Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=1, warm_start=False)

In [30]:
clf.intercept_

0.0015901396177636588

In [31]:
clf.coef_ #5, 15 26,27,29,30

array([  0.00000000e+00,  -0.00000000e+00,   0.00000000e+00,
         8.87716216e-06,   0.00000000e+00,   0.00000000e+00,
         0.00000000e+00,   0.00000000e+00,  -0.00000000e+00,
        -0.00000000e+00,   0.00000000e+00,   2.34557913e-08,
        -0.00000000e+00,   0.00000000e+00,  -0.00000000e+00,
         0.00000000e+00,  -2.29232339e-09,  -3.39350808e-09,
        -2.11106971e-09,  -2.32669211e-07,  -0.00000000e+00,
         0.00000000e+00,  -0.00000000e+00,   0.00000000e+00,
         0.00000000e+00,  -0.00000000e+00,   0.00000000e+00,
        -0.00000000e+00,   0.00000000e+00,   0.00000000e+00,
        -0.00000000e+00,   0.00000000e+00,  -0.00000000e+00,
         0.00000000e+00,   0.00000000e+00,  -0.00000000e+00,
         0.00000000e+00,  -0.00000000e+00,  -0.00000000e+00,
        -0.00000000e+00,  -0.00000000e+00,  -0.00000000e+00,
         0.00000000e+00,  -0.00000000e+00,  -0.00000000e+00,
         0.00000000e+00,  -0.00000000e+00,  -0.00000000e+00,
        -0.00000000e+00]

In [32]:
from sklearn.metrics import mean_squared_error
from math import sqrt


array_test = test.values
array_train = train.values

reg=linear_model.LinearRegression()
reg.fit(array_train[:,[5,15,26,27,29,30]], array_train[:,33])
p = reg.predict(array_test[:,[5,15,26,27,29,30]])
print(sqrt(mean_squared_error(array_test[:,33],p)))

0.15595630175181358


### Manual Selection

In [140]:

lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[3,7,11,13,15,26,29,38,39,41,42,43,45,48,50,51,52,53,54,55,56,57,58,59,60]], array_train[:,33])


#array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[3,7,11,13,15,26,29,38,39,41,42,43,45,48,50,51,52,53,54,55,56,57,58,59,60]])

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))


0.16029404835498545


All:               0.15994609372231042
Remove 2:          0.1599461468254335
Remove 3:          0.15994139947869387
Add 3, remove 4:   0.15994601407578807
Remove 5:          0.16002736613201468
Remove 7:          0.16002560995327034
Add 7, remove 8:   0.16004570567106657
Remove 9:          0.1600467659719313
Remove 10:         0.16005384969415262 
Remove 11:         0.1600511367437618
Add 11, remove 13: 0.16005327618207021
Add 13, remove 14: 0.16005644584053777 
Remove 15:         0.16004462485611465
Add 15, remove 16: 0.16007533377960984 
Remove 21:         0.16017049006374046 
Remove 22:         0.16017170517249144
Remove 24:         0.1601783899872867
Remove 25:         0.1601796287662811
Remove 26:         0.16017706982032512
Add 26, remove 27: 0.16018026912270933  
Remove 29:         0.16017920035754296
Add 29, remove 30: 0.16023854885917077       
Remove 31:         0.1602925417914655
Remove 38:         0.16029254179143937
Add 38, remove 39: 0.16029254179142366
Add 39, remove 40: 0.1602925417914472
Remove 41:         0.16029254179144642
Add 41, remove 42: 0.16029254179144642
Add 42, remove 43: 0.1602925385331169
Add 43, remove 44: 0.16029254179144858
Remove 45:         0.16029251923916113
Add 45, remove 46: 0.1602927737180532
Remove 47:         0.16029369722853465
Remove 48:         0.16029366154133506
Add 48, remove 49: 0.16029404835498545   ****
Remove 50:         0.16029224207830797
Add 50, remove 51: 0.1602921893748259
Add 51, remove 52: 0.16029202797063402
Add 52, remove 53: 0.1602927328930735
Add 53, remove 54: 0.16029339478868923
Add 54, remove 55: 0.1602940483549817
Add 55, remove 56: 0.16029255811077978
Add 56, remove 57: 0.16029366638148124
Add 57, remove 58: 0.16029246489881105
Add 58, remove 59: 0.16029228381359256
Add 59, remove 60: 0.16029145563460911


Best: 0.16029404835498545

buildingqualitytypeid
fireplacecnt
hashottuborspa
latitude
lotsizesqarefeet
structuretaxvaluedollarcnt
landtaxvaluedollarcnt
FIPS codes 6037 and 6059
Set Year
AC types None, Yes, Central (5, 13, 1)
Heat except baseboard heat (1)

In [141]:
# Manual Selection
lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[42,43]], array_train[:,33])


#array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[42,43]])

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))

0.16030219724685932


In [142]:
# Manual Selection
lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[42,43,11]], array_train[:,33])


#array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[42,43,11]])

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))

0.16030273953373883


In [144]:
# Manual Selection
lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[42,43,11,3]], array_train[:,33])


#array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[42,43,11,3]])

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))

0.160304916311392


In [146]:
# Manual Selection
lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[42,43,11,3,15]], array_train[:,33])


#array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[42,43,11,3,15]])

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))

0.16031790171710306


In [147]:
# Manual Selection
lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[42,43,11,3,15,26]], array_train[:,33])


#array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[42,43,11,3,15,26]])

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))

0.16032524933869377


In [150]:
# Manual Selection
lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[42,43,11,3,15,26,49,50,51,52,53,54,55,56,57,58,59,60]], array_train[:,33])


#array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[42,43,11,3,15,26,49,50,51,52,53,54,55,56,57,58,59,60]])

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))

0.16033215899151984


In [177]:
lm.coef_

array([ -1.53422972e-03,  -7.47628117e-03,  -3.86553879e-03,
         2.15165278e-08,   1.39227432e-08,   7.27889901e-03,
        -1.78757213e-03,   5.13507524e-02,   1.21968638e-02,
         9.61003664e-03,   8.39647764e-03,  -4.32864975e-02,
        -4.39234844e-02,   9.54097912e-18,   1.45245053e-03,
        -4.55601526e-03,   2.00155766e-02,  -2.27173589e-03,
        -8.98442452e-03])

In [178]:
lm.intercept_

0.13744687268075786

In [153]:
# Manual Selection
lm=linear_model.LinearRegression()
array_train=train.values
lm.fit(array_train[:,[3,11,13,15,26,42,43,49,50,51,52,53,54,55,56,57,58,59,60]], array_train[:,33])


#array_test = test.values

# Make predictions using the testing set
log_pred = lm.predict(test.iloc[:,[3,11,13,15,26,42,43,49,50,51,52,53,54,55,56,57,58,59,60]])

print(sqrt(mean_squared_error(array_test[:,33], log_pred)))

0.16033999669758048


In [176]:
# use cannot use MAPE because cannot divide by zero, use SMAPE instead
# https://math.stackexchange.com/questions/677852/how-to-calculate-relative-error-when-true-value-is-zero
# https://en.wikipedia.org/wiki/Symmetric_mean_absolute_percentage_error

y_true = np.array(array_test[:,33])
y_score = np.array(log_pred)

np.mean(np.abs(y_true - y_score) / (np.abs(y_true) + np.abs(y_score))) * 100

72.32086013904555

# Random Forests

In [34]:
from sklearn.model_selection import train_test_split

array=df.values
Y=array[:,33]
Y

array([0.0953, 0.0198, 0.006, ..., 0.725458273984, 0.043652112878300016,
       -0.0141783454295], dtype=object)

In [35]:
x1=array[:,2:6]
x2=array[:,7:12]
x3=array[:,13:17]
x4=array[:,21:32]  
x5=array[:,38:65]

X=np.hstack((x1,x2))
X=np.hstack((X,x3))
X=np.hstack((X,x4))
X=np.hstack((X,x5))

X[1]

array([2.0, 6.015460983643267, 1.0, 777.0, 0.0, 1.0, 1.0, 434.0, 0,
       34.272866, -119.19891100000001, 7200.0, 0.0, 4.0, 1.0, 1990.0, 1.0,
       0, 143809.0, 239679.0, 2015.0, 95870.0, 2581.3, 0, 0, 0, 1, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0], dtype=object)

In [37]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2)

* Attempt 1: Defaults
* Attempt 2: n_estimators = 20
* Attempt 3: n_estimators = 100
* Attempt 4: n_estimators = 100, max_features=20
* Attempt 5: n_estimators = 200, max_features=20
* Attempt 6: max_features=20
* Attempt 7: n_estimators=60, max_features=15
* Attempt 8: n_estimators=20, max_features=15
* Attempt 9: max_features=10, oob_score = True

In [90]:
from sklearn.ensemble import RandomForestRegressor
forest = RandomForestRegressor(n_estimators=100, max_features=10, oob_score = True)
forest.fit(X_train, Y_train)



RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features=10, max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           n_estimators=100, n_jobs=1, oob_score=True, random_state=None,
           verbose=0, warm_start=False)

In [91]:
forest.feature_importances_

array([  1.94250800e-02,   1.95743989e-02,   1.44195901e-02,
         9.03346426e-02,   3.57351055e-03,   1.14410804e-02,
         5.96119295e-03,   1.92752310e-02,   8.90184553e-04,
         8.99455276e-02,   9.08788566e-02,   8.43657887e-02,
         3.77879491e-03,   2.45748238e-02,   5.46089592e-03,
         6.81738826e-02,   4.37351104e-03,   2.78964453e-03,
         9.67743348e-02,   9.20705223e-02,   6.60451840e-03,
         9.33658908e-02,   9.55680016e-02,   4.42202228e-03,
         2.31180047e-03,   2.50363650e-03,   1.24924603e-03,
         6.91364969e-03,   6.44001999e-03,   5.00748969e-03,
         2.69348125e-07,   5.97326991e-03,   1.46510655e-04,
         2.10105173e-05,   1.31133998e-03,   1.72114142e-05,
         5.12940735e-03,   5.01793364e-04,   4.48957643e-03,
         2.79476875e-07,   6.86671854e-08,   1.30478733e-08,
         1.92597770e-05,   4.32009680e-08,   4.70156903e-05,
         2.73225329e-05,   6.81227268e-04,   3.82022845e-03,
         6.16848773e-04,

In [92]:
pf = forest.predict(X_test)

In [93]:
print(sqrt(mean_squared_error(Y_test, pf)))

0.16305236636071274


In [94]:

np.mean(np.abs(Y_test - pf) / (np.abs(Y_test) + np.abs(pf))) * 100

72.2697335021307

In [95]:
from sklearn import svm
from sklearn.model_selection import GridSearchCV

svc = svm.SVC()
clf = GridSearchCV(svc)
grid = clf.fit(X_train, Y_train)

TypeError: __init__() missing 1 required positional argument: 'param_grid'

In [108]:
from sklearn.feature_selection import RFE

forest = RandomForestRegressor(max_features=20)
rfe = RFE(forest, n_features_to_select=20)
rfe.fit(X_train,Y_train)
rfe.ranking_ 

array([ 1,  1,  1,  1, 10,  1,  6,  1, 21,  1,  1,  1,  8,  1,  1,  1,  9,
       12,  1,  1,  1,  1,  1,  3, 14, 13, 19,  1,  1,  5, 32,  4, 22, 27,
       15, 24,  7, 20,  1, 30, 29, 31, 25, 28, 23, 26, 18,  2, 17, 11, 16])

In [110]:
rfe_p = rfe.predict(X_test)

In [111]:
print(sqrt(mean_squared_error(Y_test, rfe_p)))

0.17324768609162763


In [112]:

np.mean(np.abs(Y_test - rfe_p) / (np.abs(Y_test) + np.abs(rfe_p))) * 100

73.11897961782759

In [113]:
x1=array[:,2:6]
x2=array[:,8:9]
x3=array[:,10:11]
x4=array[:,13:16]  
x5=array[:,21:24]
x6=array[:,26:31]
x7=array[:,41:43]
x8=array[:,52:53]

X=np.hstack((x1,x2))
X=np.hstack((X,x3))
X=np.hstack((X,x4))
X=np.hstack((X,x5))
X=np.hstack((X,x6))
X=np.hstack((X,x7))
X=np.hstack((X,x8))


X[1]

array([2.0, 6.015460983643267, 1.0, 777.0, 1.0, 434.0, 34.272866,
       -119.19891100000001, 7200.0, 4.0, 1.0, 1990.0, 143809.0, 239679.0,
       2015.0, 95870.0, 2581.3, 1, 0, 0], dtype=object)

In [114]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2)

In [119]:
forest = RandomForestRegressor()
forest.fit(X_train, Y_train)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

In [120]:
pf = forest.predict(X_test)

In [121]:
print(sqrt(mean_squared_error(Y_test, pf)))

0.1800865595207027
